File: OOP_C++/tema5/lab5_2.cpp
#include<iostream>
#include<cmath>
using namespace std;

class CEO{
protected:
    string nume;
    double salariu;
public:
    CEO(string n = "Jeff Bezos", double s = 10) : nume(n), salariu(s) {}
 virtual void salary(){
     cout<<salariu<<"M per year"<<endl;
   }
    virtual void name() {
     cout<<nume<<endl;
 }
};

class Manager:public CEO{
public:
    Manager(string n = "Mark Henry", double s= 100) : CEO(n, s) {}
    void salary() override{
        cout<<"Could buy a Porsche, but drives a Prius instead "<<"(around "<<salariu<<"k a year)"<<endl;
    }

    void name() override{
        cout<<nume<<endl;
    }

};

class Developer:public Manager{
    public:
    Developer(string n = "Matt", double s = 50) : Manager(n, s) {}

    void salary() override{
        cout<<"Loans a Porsche to drive around, lying to himself he will buy it someday"<<"(around "<<salariu<<"k a year)"<<endl;
    }

    void name() override{
        cout<<nume<<endl;
    }

};


int main(){
    Developer dev;
    Manager manager;
    CEO ceo;

    CEO* p1=&ceo;
    CEO* p2=&manager;
    CEO* p3=&dev;

    p1->name();
    p1->salary();

    p2->name();
    p2->salary();

    p3->name();
    p3->salary();
    return 0;
}

File: OOP_Kotlin/tema5/lab5_2.kt
open class CEO(
    open val nume: String = "Jeff Bezos",
    open val salariu: Double = 10.0
) {
    open fun salary() {
        kotlin.io.println("${salariu}M per year")
    }

    open fun name() {
        kotlin.io.println(nume)
    }
}

open class Manager(
    override val nume: String = "Mark Henry",
    override val salariu: Double = 100.0
) : CEO(nume, salariu) {
    override fun salary() {
        kotlin.io.println("Could buy a Porsche, but drives a Prius instead (around ${salariu}k a year)")
    }
}

class Developer(
    override val nume: String = "Matt",
    override val salariu: Double = 50.0
) : Manager(nume, salariu) {
    override fun salary() {
        kotlin.io.println("Loans a Porsche to drive around, lying to himself he will buy it someday (around ${salariu}k a year)")
    }
}

fun main(){
    val ceo=CEO();
    ceo.name()
    ceo.salary()


    val manager=Manager()
    manager.name()
    manager.salary()


    val developer=Developer()
    developer.name()
    developer.salary()

}